### PR TITLE
WIP Implement passport authentication [Do not Merge]

### PR DESCRIPTION
**Pull request by brainyfarm**
#### What does this PR do?
Setup passport for social authentication

#### Description of Task to be completed?
- Setup Passport
- Implement Google Login Strategy
- Implement LinkedIn Login Strategy
- Implement Facebook Login Strategy

#### How should this be manually tested?
Run the app
- For Google, visit `localhost:4001/auth/google` and login with google

#### Any background context you want to provide?
N/A

#### What are the relevant stories?
Setup and implement passport authentication

#### Screenshots (if appropriate)
Screen Shot 2019-03-15 at 2 13 43 PM

Screen Shot 2019-03-15 at 5 22 18 PM

#### Questions:
N/A

### CODE REVIEW

File: passport/LinkedIn.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { Strategy as LinkedInStrategy } from 'passport-linkedin-oauth2';
+
+import User from '../models/User';
+import config from '../config/keys';
+
+export default new LinkedInStrategy({
+  clientID: config.linkedIn.clientID,
+  clientSecret: config.linkedIn.clientSecret,
+  callbackURL: '/auth/linkedin/success'
+}, (accessToken, refresh, profile, done) => User.findOne({ linkedInId: profile.id })
```
Comment (brainyfarm) [inline passport/LinkedIn.js]: Looks rough looking down from here.

File: controllers/Auth.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable */
```
Comment (brainyfarm) [inline controllers/Auth.js]: Dude, fix these issues and don't disable the linter.

File: models/User.js [hunk 3]
```diff
@@ -1,18 +1,15 @@
...
-  email: {
-    type: String,
-    required: 'Please enter a title'
-  },
-  status: {
-    type: String,
-    enum: ['done', 'not started', 'in progress'],
-    default: 'not started'
-  }
+  name: String,
```
Comment (brainyfarm) [inline models/User.js]: These fields are too basic and require more validations and meta fields.

File: passport/Google.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { Strategy as GoogleStrategy } from 'passport-google-oauth20';
+
+import User from '../models/User';
+import config from '../config/keys';
+
+export default new GoogleStrategy({
+  clientID: config.google.clientID,
+  clientSecret: config.google.clientSecret,
+  callbackURL: '/auth/google/success',
+}, (accessToken, refresh, profile, done) => User.findOne({ googleId: profile.id })
```
Comment (brainyfarm) [inline passport/Google.js]: `findOrCreate` can do the job here since it's probably internally optimized.

File: passport/LinkedIn.js (hunk 1 above)
Comment (brainyfarm) [inline passport/LinkedIn.js]: You can probably keep this callback outside to make this thing more readable.

File: passport/LinkedIn.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  .then((isExistingUser) => {
```
Comment (brainyfarm) [inline passport/LinkedIn.js]: `isExistingUser` suggests the data type is a boolean, use something else biko.

File: .env.example [hunk 6]
```diff
@@ -0,0 +1,5 @@
+MONGOURI = mongodb://localhost/MemeMakerDB
+GoogleClientID = 8734453487-8kpp2i94werfwtewfwfwdh35osl34t.apps.googleusercontent.com
+GoogleClientSecret = jhwejhdjhdhjsdhjdhjkd
+LinkedInClientID = 89bddahvtr4ay
+LinkedInClientSecret = JVwjrep8F4G
```
Comment (Veraclins) [inline .env.example:5]: @brainyfarm Are these real keys or just samples?

File: config/keys.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    clientID: process.env.LinkedInClientID,
+    clientSecret: process.env.LinkedInClientSecret,
+  },
+  facebook: {
+    clientID: process.env.FacebookAPPID,
+    clientSecret: process.env.FacebookAPPSecret,
+  },
+  mongodb: {
+    dbURI: process.env.MONGOURI,
+  }
```
Comment (Veraclins) [inline config/keys.js]: @brainyfarm for consistency and to keep with convention, environment variables should be all caps with underscores, eg, `FACEBOOK_CLIENT_ID`

File: passport/Google.js (hunk 4 above)
Comment (Veraclins) [reply passport/Google.js]: I agree with using `findOrCreate`, though I don't know if mongoose has findOrCreate

File: passport/LinkedIn.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        name: profile.displayName,
+        email: profile.emails[0].value,
+        linkedInId: profile.id,
+        imageUrl: profile._json.pictureUrl,
+      });
+      return newUser.save()
+        .then((user) => {
+          done(null, user);
+        });
+    }
```
Comment (Veraclins) [inline passport/LinkedIn.js]: you may consider abstracting some of the repeated codes into a reusable function. Also, I think it is necessary to differentiate new user creation from existing user verification (signup vs sign in) as that information may be important on the frontend

File: routes/auth.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+router.route('/facebook')
+  .get(passport.authenticate('facebook', {
+    scope: [
+      'read_stream',
+      'r_basicprofile'
+    ]
+  }));
+
+router.route('/facebook/success')
+  .get(passport.authenticate('facebook', { session: false }), Auth.loginSuccess);
```
Comment (Veraclins) [inline routes/auth.js]: You can abstract this to prevent duplication

**Review by Veraclins — CHANGES_REQUESTED** [review]
Great job @brainyfarm, I left some comments below.

**Pull request closed without merging**